Incorrect translation of negative constraints

Hi, I'm a maintainer of Homebrew. We're looking to use `pipgrip` as our dependency resolver to pin our Python packages to specific dependency trees. We've found that certain negative constraints such as in [cryptography](https://github.com/pyca/cryptography/blob/master/pyproject.toml) are not correctly translated during dependency resolution. In particular, `cffi>=1.8,!=1.11.3` turns into `cffi<1.11.3|>1.11.3,>=1.8`, which I think due to some order of operations issues, `pip` gets confused and selects the broken 1.11.3 build anyway.

If I use the Python debugger and replace the bad constraint with the proper negative constraint `cffi>=1.8,!=1.11.3`, or specify this constraint on the command line, the build finishes fine. 

<details><summary>Correct build (with extra constraint)</summary>

```console
% pipgrip --no-cache-dir jrnl==2.4.4 'cffi>=1.8,!=1.11.3'
jrnl==2.4.4
ansiwrap==0.8.4
textwrap3==0.9.2
asteval==0.9.18
colorama==0.4.3
cryptography==2.9.2
cffi==1.14.0
pycparser==2.20
six==1.15.0
keyring==21.2.1
parsedatetime==2.6
passlib==1.7.2
python-dateutil==2.8.1
pytz==2020.1
pyxdg==0.26
pyyaml==5.3.1
tzlocal==2.1
```

</details>

<details><summary>Crashing build</summary>

```console
% pipgrip --no-cache-dir jrnl==2.4.4 -vv         
INFO: fact: _root_ is root
INFO: derived: root
INFO: fact: root depends on jrnl (==2.4.4)
INFO: selecting . (0.0.0)
INFO: derived: jrnl (==2.4.4)
INFO: fact: jrnl (2.4.4) depends on ansiwrap (<0.9.0,>=0.8.4)
INFO: fact: jrnl (2.4.4) depends on asteval (>=0.9.14,<0.10.0)
INFO: fact: jrnl (2.4.4) depends on colorama (>=0.4.1,<0.5.0)
INFO: fact: jrnl (2.4.4) depends on cryptography (<3.0,>=2.7)
INFO: fact: jrnl (2.4.4) depends on keyring (<22.0,>19.0)
INFO: fact: jrnl (2.4.4) depends on parsedatetime (>=2.4,<3.0)
INFO: fact: jrnl (2.4.4) depends on passlib (<2.0,>=1.7)
INFO: fact: jrnl (2.4.4) depends on python-dateutil (<3.0,>=2.8)
INFO: fact: jrnl (2.4.4) depends on pytz (>=2019.1,<2021.0)
INFO: fact: jrnl (2.4.4) depends on pyxdg (<0.27.0,>=0.26.0)
INFO: fact: jrnl (2.4.4) depends on pyyaml (<6.0,>=5.1)
INFO: fact: jrnl (2.4.4) depends on tzlocal (<3.0,>1.5)
INFO: selecting jrnl (2.4.4)
INFO: derived: tzlocal (<3.0,>1.5)
INFO: derived: pyyaml (<6.0,>=5.1)
INFO: derived: pyxdg (<0.27.0,>=0.26.0)
INFO: derived: pytz (>=2019.1,<2021.0)
INFO: derived: python-dateutil (<3.0,>=2.8)
INFO: derived: passlib (<2.0,>=1.7)
INFO: derived: parsedatetime (>=2.4,<3.0)
INFO: derived: keyring (<22.0,>19.0)
INFO: derived: cryptography (<3.0,>=2.7)
INFO: derived: colorama (>=0.4.1,<0.5.0)
INFO: derived: asteval (>=0.9.14,<0.10.0)
INFO: derived: ansiwrap (<0.9.0,>=0.8.4)
INFO: selecting pyxdg (0.26)
INFO: fact: ansiwrap (0.8.4) depends on textwrap3 (>=0.9.2)
INFO: selecting ansiwrap (0.8.4)
INFO: derived: textwrap3 (>=0.9.2)
INFO: selecting textwrap3 (0.9.2)
INFO: fact: python-dateutil (2.8.1) depends on six (>=1.5)
INFO: selecting python-dateutil (2.8.1)
INFO: derived: six (>=1.5)
INFO: selecting passlib (1.7.2)
INFO: selecting parsedatetime (2.6)
INFO: selecting colorama (0.4.3)
INFO: fact: tzlocal (2.1) depends on pytz (*)
INFO: selecting tzlocal (2.1)
INFO: selecting pytz (2020.1)
INFO: selecting asteval (0.9.18)
INFO: fact: cryptography (2.9.2) depends on cffi (>=1.8,<1.11.3 || >1.11.3)
INFO: fact: cryptography (2.9.2) depends on six (>=1.4.1)
INFO: selecting cryptography (2.9.2)
INFO: derived: cffi (>=1.8,<1.11.3 || >1.11.3)
ERROR: Collecting cffi<1.11.3|>1.11.3,>=1.8
  Using cached cffi-1.11.3.tar.gz (436 kB)
Building wheels for collected packages: cffi
  Building wheel for cffi (setup.py): started
  Building wheel for cffi (setup.py): finished with status 'error'
  ERROR: Command errored out with exit status 1:
   command: /usr/local/Cellar/pipgrip/0.4.0/libexec/bin/python3.8 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/sy/w20_4b6s2ls83kqdtdtskd3r0000gn/T/pip-wheel-f6myglmi/cffi/setup.py'"'"'; __file__='"'"'/private/var/folders/sy/w20_4b6s2ls83kqdtdtskd3r0000gn/T/pip-wheel-f6myglmi/cffi/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /private/var/folders/sy/w20_4b6s2ls83kqdtdtskd3r0000gn/T/pip-wheel-vbizfeii
       cwd: /private/var/folders/sy/w20_4b6s2ls83kqdtdtskd3r0000gn/T/pip-wheel-f6myglmi/cffi/
  Complete output (62 lines):
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.macosx-10.15-x86_64-3.8
  creating build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/backend_ctypes.py -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/error.py -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/setuptools_ext.py -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/__init__.py -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/cffi_opcode.py -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/vengine_gen.py -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/model.py -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/ffiplatform.py -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/api.py -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/vengine_cpy.py -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/commontypes.py -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/lock.py -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/recompiler.py -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/cparser.py -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/verifier.py -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/_cffi_include.h -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/parse_c_type.h -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/_embedding.h -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/_cffi_errors.h -> build/lib.macosx-10.15-x86_64-3.8/cffi
  running build_ext
  building '_cffi_backend' extension
  creating build/temp.macosx-10.15-x86_64-3.8
  creating build/temp.macosx-10.15-x86_64-3.8/c
  clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk -I/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include -I/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Tk.framework/Versions/8.5/Headers -DUSE__THREAD -DHAVE_SYNC_SYNCHRONIZE -I/usr/local/Cellar/libffi/3.3/include -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/ffi -I/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/include/python3.8 -c c/_cffi_backend.c -o build/temp.macosx-10.15-x86_64-3.8/c/_cffi_backend.o
  c/_cffi_backend.c:5827:9: warning: 'ffi_prep_closure' is deprecated [-Wdeprecated-declarations]
      if (ffi_prep_closure(closure, &cif_descr->cif,
          ^
  /usr/local/Cellar/libffi/3.3/include/ffi.h:341:18: note: 'ffi_prep_closure' has been explicitly marked deprecated here
    __attribute__((deprecated))
                   ^
  In file included from c/_cffi_backend.c:7286:
  In file included from c/cffi1_module.c:20:
  c/call_python.c:20:30: error: incomplete definition of type 'struct _is'
      builtins = tstate->interp->builtins;
                 ~~~~~~~~~~~~~~^
  /usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/include/python3.8/pystate.h:20:8: note: forward declaration of 'struct _is'
  struct _is;
         ^
  In file included from c/_cffi_backend.c:7286:
  In file included from c/cffi1_module.c:20:
  c/call_python.c:160:39: error: incomplete definition of type 'struct _is'
      new1 = PyThreadState_GET()->interp->modules;
             ~~~~~~~~~~~~~~~~~~~~~~~~~~~^
  /usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/include/python3.8/pystate.h:20:8: note: forward declaration of 'struct _is'
  struct _is;
         ^
  In file included from c/_cffi_backend.c:7286:
  In file included from c/cffi1_module.c:20:
  c/call_python.c:249:63: error: incomplete definition of type 'struct _is'
          if (externpy->reserved1 != PyThreadState_GET()->interp->modules) {
                                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~^
  /usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/include/python3.8/pystate.h:20:8: note: forward declaration of 'struct _is'
  struct _is;
         ^
  1 warning and 3 errors generated.
  error: command 'clang' failed with exit status 1
  ----------------------------------------
  ERROR: Failed building wheel for cffi
  Running setup.py clean for cffi
Failed to build cffi
ERROR: Failed to build one or more wheels

Error: Command '['/usr/local/Cellar/pipgrip/0.4.0/libexec/bin/python3.8', '-m', 'pip', 'wheel', '--no-deps', '--disable-pip-version-check', '--wheel-dir', '/var/folders/sy/w20_4b6s2ls83kqdtdtskd3r0000gn/T/tmpxcz0ybav', '--index-url', 'https://pypi.org/simple', '--trusted-host', 'pypi.org', '--progress-bar=off', 'cffi<1.11.3|>1.11.3,>=1.8']' returned non-zero exit status 1.
```

</details>

_Originally posted by @jonchang in https://github.com/Homebrew/homebrew-core/pull/58626#issuecomment-663932846_

--- a/src/pipgrip/package_source.py
+++ b/src/pipgrip/package_source.py
@@ -155,12 +155,9 @@ class PackageSource(BasePackageSource):
 
         """
         extras = package.req.extras
-        version_specifier = str(constraint).replace("||", "|").replace(" ", "")
-        if version_specifier and version_specifier[0].isdigit():
-            version_specifier = "==" + version_specifier
-        req = parse_req(str(package) + version_specifier, extras)
         if package not in self._packages or extras not in self._packages[package]:
-            self.discover_and_add(req.__str__())
+            # unseen package, safe to take initially parsed req directly
+            self.discover_and_add(package.req.__str__())
         if package not in self._packages:
             return []
 

--- a/tests/test_cli.py
+++ b/tests/test_cli.py
@@ -21,7 +21,7 @@ def mock_download_wheel(package, *args, **kwargs):
         "wheel": "./tests/assets/wheel-0.33.6-py2.py3-none-any.whl",
         "pyparsing>=2.0.2": "./tests/assets/pyparsing-2.4.6-py2.py3-none-any.whl",
         "requests==2.22.0": "./tests/assets/requests-2.22.0-py2.py3-none-any.whl",
-        "urllib3<1.25.0|>1.25.0,<1.25.1|>1.25.1,<1.26,>=1.21.1": "./tests/assets/urllib3-1.25-py2.py3-none-any.whl",
+        "urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1": "./tests/assets/urllib3-1.25.7-py2.py3-none-any.whl",
         "urllib3==1.25.7": "./tests/assets/urllib3-1.25.7-py2.py3-none-any.whl",
         "idna<2.9,>=2.5": "./tests/assets/idna-2.8-py2.py3-none-any.whl",
         "chardet<3.1.0,>=3.0.2": "./tests/assets/chardet-3.0.4-py2.py3-none-any.whl",
@@ -68,6 +68,10 @@ def mock_get_available_versions(package, *args, **kwargs):
 # fmt: on
 
 
+def mock_stream_bash_command(*args, **kwargs):
+    return "I passed"
+
+
 def invoke_patched(func, arguments, monkeypatch):
     def default_environment():
         return {
@@ -92,6 +96,9 @@ def invoke_patched(func, arguments, monkeypatch):
     )
     monkeypatch.setattr(
         pipgrip.pipper, "default_environment", default_environment,
+    )
+    monkeypatch.setattr(
+        pipgrip.pipper, "stream_bash_command", mock_stream_bash_command,
     )
     runner = CliRunner()
     return runner.invoke(main, arguments)
@@ -221,7 +228,7 @@ def test_solutions(arguments, expected, monkeypatch):
         ),
         (["-r", "tests/test_reqs.txt"]),
         (["-r", "tests/test_reqs.txt", "-r", "tests/test_reqs.txt"]),
-        (["urllib3==1.25.7", "urllib3<1.25.0|>1.25.0,<1.25.1|>1.25.1,<1.26,>=1.21.1"]),
+        (["urllib3==1.25.7", "urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1"]),
         # (["keras-preprocessing==1.0.2", "keras==2.2.2"]),  # fix RecursionError
     ],
 )

